Some fixes following your suggestions

beepsoft opened the pull request:
Thank you very much for the repo and the details comments, suggestions and updates!

Here are my changes, also added to the readme.

# Changes 2023.02.06

I completely agree with your suggestions and changes.

## Fixed use of compiled style.css in the web component

In [index.wc.js](src%2Fcrate-builder%2Findex.wc.js) I added the compiled `style.css` to the styles array of the DescriboCrateBuilder builder, this way the style is avaialble inside the web component.

## Example of passing non-string values to the web component

My current solution for passing objects and functions to the web component is to do it via [globalThis](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/globalThis), see [Shell.component.wc.vue](src%2Fcrate-builder%2FShell.component.wc.vue). Users of the web component should pass the name of the function/object as it is set on globalThis (and the users should take care of avoiding name clashes). See [App.tsx](wc-test-app-react%2Fsrc%2FApp.tsx).

From now on translation of web component properties can happen in [Shell.component.wc.vue](src%2Fcrate-builder%2FShell.component.wc.vue) and then passed to [Shell.component.vue](src%2Fcrate-builder%2FShell.component.vue) in the expected format.

beepsoft committed:
Use describo-crate-builder-wc.umd from parent build

beepsoft committed:
Fix inclusion of css in web components

There might be a more straightforward way using but I couldn't get it working:
https://vuejs.org/guide/extras/web-components.html#building-custom-elements-with-vue

beepsoft committed:
Add example for passing functions/object from react to the web component

beepsoft committed:
Update readme with changes

marcolarosa committed:
add commentary / question about including the style here

marcolarosa committed:
refactor to include all component config via globalThis

marcolarosa commented:
hi @beepsoft 

This looks good. I've taken your idea and extended it to all of the configuration for the component. Have a look at https://github.com/describo/describo-web-component-test/commit/450d1a32c6eb88ee825660e038434aece2bba156 and let me know how you feel about this. In this way there's no need to stringify/parse data. The environment (react in this case) can just prepare the state and then instantiate the web component.

It's worth noting that by using globalThis there is the possibility that the variable name will clash if the component is used more than once in the app but I expect the likelihood of that is tiny. It shouldn't clash with anything else as I've named it [globalThis.DescriboCrateBuilderConfiguration](https://github.com/describo/describo-web-component-test/commit/450d1a32c6eb88ee825660e038434aece2bba156#diff-307432f39ffbbe7adff7efd67440b7c63a3cdd917415ed7bc2cd81d773f6c159R53).

Also, I have a question to you about importing the style into the component in [index.wc.js](https://github.com/describo/describo-web-component-test/commit/450d1a32c6eb88ee825660e038434aece2bba156#diff-307432f39ffbbe7adff7efd67440b7c63a3cdd917415ed7bc2cd81d773f6c159R53).

The other thing we need to sort out if events coming out of the web component. The Describo component emits events to tell the outer app (react in this case) that data needs to be saved or the app is ready: https://github.com/describo/crate-builder-component#events. So this component needs some event handlers to received those events from the vue component and then send them out to the calling environment.

beepsoft committed:
Remove style.css import, pass config via prop

beepsoft commented:
Hi @marcolarosa,

I made the following changes in https://github.com/describo/describo-web-component-test/pull/1/commits/e3f60337784dce37abfca64c98aefbb749532406, PR https://github.com/describo/describo-web-component-test/pull/2 (based on your [wc-trial](https://github.com/describo/describo-web-component-test/tree/wc-trial) branch)

1. Removed import of `style.css` in react [App.tsx](wc-test-app-react%2Fsrc%2FApp.tsx) as it is not required.

3. Instead of accessing `globalThis.DescriboCrateBuilderConfiguration` in [Shell.component.wc.vue](src%2Fcrate-builder%2FShell.component.wc.vue) I suggest to pass the name of the config object on `globalThis` via the `config` prop. This way we could theoretically have multiple components on screen with different configs, and it is the responsibility of the caller to avoid name clashes.

4. The React side is now reactive: when you press the "Change crate" button the value of the crate field changes in `globalThis[props.config]`. However, this change is not picked up in [Shell.component.wc.vue](src%2Fcrate-builder%2FShell.component.wc.vue). Can you please take a look at this? 

Receiving events like `save:crate` already worked in the original PR, so that should be OK.

marcolarosa committed:
tinkering with making globalThis reactive inside vue

beepsoft committed:
Rename configVersion to configversion

Web component properties must be in lowercase. With this now all changes
from react are propagated to Describo, but the code in watch() is
a duplicate, there must be a better way to do the updates.

marcolarosa commented:
superseded by [1f6338777ae210e8934cf4a4ea802b45fa076771](https://github.com/describo/crate-builder-component/commit/1f6338777ae210e8934cf4a4ea802b45fa076771)
